### PR TITLE
8337111: Bad HTML checker for generated documentation

### PR DESCRIPTION
**Pull request by nizarbenalla**
Can I please get a review for this PR that adds 4 new html "Checkers" for the generated documentation.
More details are in the JBS issues

These tests were mostly inspired /converted from the existing [Doccheck](https://github.com/openjdk/doccheck).

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8337111](https://bugs.openjdk.org/browse/JDK-8337111): Bad HTML checker for generated documentation (**Sub-task** - P4)
 * [JDK-8337113](https://bugs.openjdk.org/browse/JDK-8337113): Bad character checker for generated documentation (**Sub-task** - P4)
 * [JDK-8337114](https://bugs.openjdk.org/browse/JDK-8337114): DocType checker for generated documentation (**Sub-task** - P4)
 * [JDK-8337116](https://bugs.openjdk.org/browse/JDK-8337116): Internal links checker for generated documentation (**Sub-task** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20711/head:pull/20711` \
`$ git checkout pull/20711`

Update a local copy of the PR: \
`$ git checkout pull/20711` \
`$ git pull https://git.openjdk.org/jdk.git pull/20711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20711`

View PR using the GUI difftool: \
`$ git pr show -t 20711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20711.diff">https://git.openjdk.org/jdk/pull/20711.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20711#issuecomment-2309728932)

### CODE REVIEW

**Comment by nizarbenalla** [comment]
Sorry for the title change.

File: test/langtools/jdk/javadoc/doccheck/DocCheck.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doccheck/DocCheck.java:43]: It is often common practice to put Java SE and JDK imports first.

File: test/langtools/jdk/javadoc/doccheck/DocCheck.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    @Test
+    public void test() throws Exception {
+        ExecutorService executorService = Executors.newFixedThreadPool(4);
+        List<Throwable> exceptions = new ArrayList<>();
+
+        try (
+                TidyChecker tidy = new TidyChecker();
+                BadCharacterChecker badChars = new BadCharacterChecker();
+                HtmlFileChecker docChecker = new HtmlFileChecker(new DocTypeChecker());
+                HtmlFileChecker htmlChecker = new HtmlFileChecker(new LinkChecker());
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doccheck/DocCheck.java:71]: I suggest extracting these declarations into a method called something like `getCheckers()`
and then having two alternative methods to call them, called homelike like `runSerially(List<Checker> checkers)` and `runWithExecutor(List<Checker> checkers)`

**Review by jonathan-gibbons — COMMENTED** [review]
Good start!  Some suggestions inline

**Comment by nizarbenalla** [comment]
I tried to address some of the review comments, I now use `TestRunner`.
I've renamed the `tools` directory to `doccheckutils` to avoid any naming conflicts.
There are now two modes of running the test, one with executors and one sequentially, you can pass a `doccheck.runParallel` command line option to test and the default value is true.

**Comment by nizarbenalla** [comment]
Closing this for now, I need time to adapt parts of this

**Pull request closed without merging**